content/blog/2019-10-23-tail-call-elimination.md

```diff
@@ -0,0 +1,239 @@
+would be used in a `push`. Additionally, a `retval` would need to be created
+afterwards if in the AST, the function call was assigned to a variable.
+
+### Identifying and Eliminating Tail Calls
+The simple definition of a *tail call* would be an immediate return of a call
+to a function. The translation from the TypeScript frontend of something like
+```
+return foo(n)
+```
+to Bril would be
+```
+push n
+call foo
+v: int = retval;
+ret v
+```
+Thus we just need to look for `call`s that are immediately and optionally
+followed by `reval`, and immediately followed by a `ret`.
+
+This doesn't take into account more complex cases where there isn't an explicit
+return of a function call, but the value returned only comes from the same
+created. For example:
+```
+function foo(n: number): number {
+  ...
+  if (b) {
+    result = foo(n-1);
+  } else {
+    result = foo(n-2);
+  }
+  return result;
+}
+```
+
+To do this, we first do a global copy propagation. The dataflow analysis for
+copy propagation that I used can be found [here](http://www.csd.uwo.ca/~moreno/CS447/Lectures/CodeOptimization.html/node8.html).
+Then, for a value `v` that is `return`ed, we search backwards through the CFG until we find a `retval` that
+corresponds to `v`, and make sure that it has not been modified along any of
+these backwards paths, and the `reval` comes from a call to the same function.
+In the above example, the corresponding Bril code looks something as follows:
+```
+then.6:
+  ...
+  call foo;
+  v17: int = retval ;
+  result: int = id v17;
+  jmp endif.6;
+else.6:
+  ...
+  call foo;
+  v21: int = retval ;
+  result: int = id v21;
+endif.6:
+  v22: int = id result;
+  ret v22;
+```
+After copy propagation, it looks like this:
+```
+then.6:
+  ...
+  call foo;
+  result: int = retval;
+  jmp endif.6;
+else.6:
+  ...
+  call foo;
+  result: int = retval;
+endif.6:
+  ret result;
+```
+
+Then we can analyze the CFG backwards to see that indeed we can replace
+the calls to a `jmp`.
+```
+then.6:
+  ...
+  jmp foo;
+  jmp endif.6;
+else.6:
+  ...
+  jmp foo;
+endif.6:
+  ret result;
+```
+Note that the extra instructions can simply be removed by a DCE pass, so we don't
+worry about that.
+
+*Unfortunately, I couldn't get this to work properly because my copy propagation
+pass had bugs.*
+
+## Evaluation
+To evaluate that the tail call elimination is working and actually gives us an
+improvement, we benchmark some recursive functions that use tail recursion, and
+show the difference in execution time and memory usage between an optimized and
+unoptimized Bril program.
+
+The table entries show how much change was observed, as a percentage, by doing
+tail call elimination (TCE). For example, an entry of -10% means the optimized program
+used 10% less memory/time than the unoptimized program. An `X` means that the output of the program was too
+big to handle. `n` is the argument passed to the recursive function.
```

Comment:
Where does this limit originate from? (Did the interpreter crash?)

Reply:
For the loop version of the program, it runs out of memory:
FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory